presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -464,123 +378,275 @@ else if (!result.getBlocked().isDone()) {
+            // never be a non-failed SectionExecution for a section that's ready for execution
+            verify(attempts.isEmpty() || getLast(attempts).isFailed(), "Non-failed sectionExecutions already exists");
+
+            PlanFragment sectionRootFragment = section.getPlan().getFragment();
+
+            Optional<int[]> bucketToPartition;
+            OutputBuffers outputBuffers;
+            ExchangeLocationsConsumer locationsConsumer;
+            if (isRootFragment(sectionRootFragment)) {
+                bucketToPartition = Optional.of(new int[1]);
+                outputBuffers = createInitialEmptyOutputBuffers(sectionRootFragment.getPartitioningScheme().getPartitioning().getHandle())
+                        .withBuffer(new OutputBufferId(0), BROADCAST_PARTITION_ID)
+                        .withNoMoreBufferIds();
+                OutputBufferId rootBufferId = getOnlyElement(outputBuffers.getBuffers().keySet());
+                locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) ->
+                        updateQueryOutputLocations(queryStateMachine, rootBufferId, tasks, noMoreExchangeLocations);
+            }
+            else {
+                bucketToPartition = Optional.empty();
+                outputBuffers = createDiscardingOutputBuffers();
+                locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) -> {};
+            }
+
+            int attemptId = attempts.size();
+            SectionExecution sectionExecution = sectionStageExecutionFactory.createSectionStageExecutions(
+                    session,
+                    section,
+                    locationsConsumer,
+                    bucketToPartition,
+                    outputBuffers,
+                    summarizeTaskInfo,
+                    remoteTaskFactory,
+                    splitSourceFactory,
+                    attemptId);
+
+            addStateChangeListeners(sectionExecution);
+            attempts.add(sectionExecution);
+            result.add(sectionExecution);
+        }
+        return result.build();
     }
 
-    private StageId getStageId(PlanFragmentId fragmentId)
+    private static void updateQueryOutputLocations(QueryStateMachine queryStateMachine, OutputBufferId rootBufferId, Set<RemoteTask> tasks, boolean noMoreExchangeLocations)
+    {
+        Map<URI, TaskId> bufferLocations = tasks.stream()
+                .collect(toImmutableMap(
+                        task -> getBufferLocation(task, rootBufferId),
+                        RemoteTask::getTaskId));
+        queryStateMachine.updateOutputLocations(bufferLocations, noMoreExchangeLocations);
+    }
+
+    private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootBufferId)
+    {
+        URI location = remoteTask.getTaskStatus().getSelf();
+        return uriBuilderFrom(location).appendPath("results").appendPath(rootBufferId.toString()).build();
```

Comment:
cc @shixuan-fan who has been working on not using URI to make coordinator latency lower.